coderd/coderdtest/oidctest/idp.go

```diff
@@ -0,0 +1,768 @@
+	t.Helper()
+
+	if _, ok := claims["exp"]; !ok {
+		claims["exp"] = time.Now().Add(time.Hour).UnixMilli()
+	}
+
+	if _, ok := claims["aud"]; !ok {
+		claims["aud"] = f.clientID
+	}
+
+	if _, ok := claims["iss"]; !ok {
+		claims["iss"] = f.issuer
+	}
+
+	signed, err := jwt.NewWithClaims(jwt.SigningMethodRS256, claims).SignedString(f.key)
+	require.NoError(t, err)
+
+	return signed
+}
+
+// httpHandler is the IDP http server.
+func (f *FakeIDP) httpHandler(t testing.TB) http.Handler {
+	t.Helper()
+
+	mux := chi.NewMux()
+	// This endpoint is required to initialize the OIDC provider.
+	// It is used to get the OIDC configuration.
+	mux.Get("/.well-known/openid-configuration", func(rw http.ResponseWriter, r *http.Request) {
+		_ = json.NewEncoder(rw).Encode(f.provider)
+	})
+
+	// Authorize is called when the user is redirected to the IDP to login.
+	// This is the browser hitting the IDP and the user logging into Google or
+	// w/e and clicking "Allow". They will be redirected back to the redirect
+	// when this is done.
+	mux.Handle(authorizePath, http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+		f.logger.Info(r.Context(), "HTTP Call Authorize", slog.F("url", string(r.URL.String())))
+
+		clientID := r.URL.Query().Get("client_id")
+		if clientID != f.clientID {
+			t.Errorf("unexpected client_id %q", clientID)
+			http.Error(rw, "invalid client_id", http.StatusBadRequest)
+		}
+
+		redirectURI := r.URL.Query().Get("redirect_uri")
+		state := r.URL.Query().Get("state")
+
+		scope := r.URL.Query().Get("scope")
+		_ = scope
+
+		responseType := r.URL.Query().Get("response_type")
+		switch responseType {
+		case "code":
+		case "token":
+			t.Errorf("response_type %q not supported", responseType)
+			http.Error(rw, "invalid response_type", http.StatusBadRequest)
+			return
+		default:
+			t.Errorf("unexpected response_type %q", responseType)
+			http.Error(rw, "invalid response_type", http.StatusBadRequest)
+			return
+		}
+
+		ru, err := url.Parse(redirectURI)
+		if err != nil {
+			t.Errorf("invalid redirect_uri %q", redirectURI)
+			http.Error(rw, "invalid redirect_uri", http.StatusBadRequest)
+			return
+		}
+
+		q := ru.Query()
+		q.Set("state", state)
+		q.Set("code", f.newCode(state))
+		ru.RawQuery = q.Encode()
+
+		http.Redirect(rw, r, ru.String(), http.StatusTemporaryRedirect)
+	}))
+
+	mux.Handle(tokenPath, http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+		values, err := f.authenticateOIDClientRequest(t, r)
+		f.logger.Info(r.Context(), "HTTP Call Token",
+			slog.Error(err),
+			slog.F("values", values.Encode()),
+		)
+		if err != nil {
+			http.Error(rw, fmt.Sprintf("invalid token request: %s", err.Error()), http.StatusBadRequest)
+			return
+		}
+		getEmail := func(claims jwt.MapClaims) string {
+			email, ok := claims["email"]
+			if !ok {
+				return "unknown"
+			}
+			if _, ok := email.(string); !ok {
+				return "wrong-type"
+			}
```

Comment:
I don't want this to fail because this is intentional sometimes.

I have hooks that use `email` so you can make a unit test that has more than 1 user and behaves differently based on that user. However there exists unit tests where the `email` field is intentionally not a string. Like sending an integer instead. 

This is to test how our oauth httpmw would handle unexpected type.

---

So basically this `getEmail` thing is just used internally with the IDP to help `hooks`, but serves no functional purpose to testing functionality.